File: src/main/java/com/HOIIVUtils/clauzewitz/data/units/SubUnit.java

package com.HOIIVUtils.clauzewitz.data.units;

import com.HOIIVUtils.clauzewitz.BoolType;
import com.HOIIVUtils.clausewitz_parser.Node;
import com.HOIIVUtils.clausewitz_parser.Parser;
import com.HOIIVUtils.clausewitz_parser.ParserException;
import com.HOIIVUtils.Settings;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

/*
 * Unit File
 */
public record SubUnit(
		String identifier,
		String abbreviation,
		String sprite,
		String mapIconCategory,
		Integer priority,
		Integer aiPriority,
		boolean active,
		String group,
		Integer combatWidth,
		Integer manpower,
		Integer maxOrganization,
		Integer defaultMorale,
		Integer maxStrength,
		Integer trainingTime,
		Double weight,
		Double supplyConsumption
) {
	public static List<SubUnit> read(File dir) {
		if (!dir.isDirectory()) {
			throw new IllegalArgumentException(dir + " is not a directory");
		}
		if (dir.listFiles().length == 0) {
			throw new IllegalArgumentException(dir + " is empty");
		}

		List<SubUnit> subUnits = new ArrayList<>();
		for (File f : dir.listFiles()) {
			if (f.isDirectory()) continue;
			Parser parser = new Parser(f);
			Node rootNode;
			try {
				rootNode = parser.parse();
			} catch (ParserException e) {
				throw new RuntimeException(e);
			}

			var l = rootNode.findFirst("sub_units").toList();
			if (l.isEmpty()) {
				System.out.println("No sub_units found in " + f.getName());
				continue;
			}
			if (Settings.DEV_MODE.enabled()) {
				System.out.println("File: " + f.getName() + ", subunits: " + l.size());
			}

			// loop through each sub unit definition in this file
			for (Node subUnitNode : l) {
//				System.out.println("subUnitNode: " + subUnitNode.name());
				SubUnit subUnit = new SubUnit(
						subUnitNode.name(),
						subUnitNode.getValue("abbreviation").string(),
						subUnitNode.getValue("sprite").string(),
						subUnitNode.getValue("map_icon_category").string(),
						subUnitNode.getValue("priority").intClass(),
						subUnitNode.getValue("ai_priority").intClass(),
						subUnitNode.getValue("active").bool(BoolType.YES_NO),
						//subUnit.type = subUnitNode.getValue("type").string(),
						subUnitNode.getValue("group").string(),
						//subUnit.categories
						subUnitNode.getValue("combat_width").intClass(),
						//subUnit.need
						subUnitNode.getValue("manpower").intClass(),
						subUnitNode.getValue("max_organization").intClass(),
						subUnitNode.getValue("default_morale").intClass(),
						subUnitNode.getValue("max_strength").intClass(),
						subUnitNode.getValue("training_time").intClass(),
						subUnitNode.getValue("weight").doubleClass(),
						subUnitNode.getValue("supply_consumption").doubleClass()
				);

				subUnits.add(subUnit);
			}
		}
		return subUnits;
	}

	public static List<Function<SubUnit, ?>> getDataFunctions() {
		List<Function<SubUnit, ?>> dataFunctions = new ArrayList<>(16);

		dataFunctions.add(SubUnit::identifier);
		dataFunctions.add(SubUnit::abbreviation);
		dataFunctions.add(SubUnit::sprite);
		dataFunctions.add(SubUnit::mapIconCategory);
		dataFunctions.add(SubUnit::priority);
		dataFunctions.add(SubUnit::aiPriority);
		dataFunctions.add(SubUnit::active);
		dataFunctions.add(SubUnit::group);
		dataFunctions.add(SubUnit::combatWidth);
		dataFunctions.add(SubUnit::manpower);
		dataFunctions.add(SubUnit::maxOrganization);
		dataFunctions.add(SubUnit::defaultMorale);
		dataFunctions.add(SubUnit::maxStrength);
		dataFunctions.add(SubUnit::trainingTime);
		dataFunctions.add(SubUnit::weight);
		dataFunctions.add(SubUnit::supplyConsumption);
		return dataFunctions;
	}

	public static List<String> getDataLabels() {
		List<String> dataFunctionLabels = new ArrayList<>(16);

		dataFunctionLabels.add("Subunit");
		dataFunctionLabels.add("Abbreviation");
		dataFunctionLabels.add("Sprite");
		dataFunctionLabels.add("Map Icon Category");
		dataFunctionLabels.add("Priority");
		dataFunctionLabels.add("AI Priority");
		dataFunctionLabels.add("Active");
		dataFunctionLabels.add("Group");
		dataFunctionLabels.add("Combat Width");
		dataFunctionLabels.add("Manpower");
		dataFunctionLabels.add("Max Organization");
		dataFunctionLabels.add("Default Morale");
		dataFunctionLabels.add("Max Strength");
		dataFunctionLabels.add("Training Time");
		dataFunctionLabels.add("Weight");
		dataFunctionLabels.add("Supply Consumption");
        return dataFunctionLabels;
    }
}

File: src/main/scala/com/hoi4utils/hoi4/units/SubUnit.scala
package com.hoi4utils.hoi4.units

import com.hoi4utils.BoolType
import com.hoi4utils.parser.{Node, Parser, ParserException}

import java.io.File
import javax.swing.JOptionPane
import scala.collection.mutable.ListBuffer
import scala.jdk.CollectionConverters.*

/**
 * A single sub‑unit definition in HOI4 (e.g. combat battalion or support company).
 *
 * @param identifier        The script ID of the sub‑unit
 * @param abbreviation      Abbreviation string (if present)
 * @param sprite            Path to sprite (if present)
 * @param mapIconCategory   Map icon category (if present)
 * @param priority          Spawn priority (if present)
 * @param aiPriority        AI priority (if present)
 * @param active            Whether this sub‑unit is active (if present)
 * @param group             Group identifier (if present)
 * @param combatWidth       Combat width (if present)
 * @param manpower          Manpower cost (if present)
 * @param maxOrganization   Maximum organization (if present)
 * @param defaultMorale     Default morale (if present)
 * @param maxStrength       Maximum strength (if present)
 * @param trainingTime      Training time in days (if present)
 * @param weight            Unit weight (if present)
 * @param supplyConsumption Supply consumption (if present)
 */
case class SubUnit(
                    identifier: String,
                    abbreviation: Option[String],
                    sprite: Option[String],
                    mapIconCategory: Option[String],
                    priority: Option[Int],
                    aiPriority: Option[Int],
                    active: Option[Boolean],
                    group: Option[String],
                    combatWidth: Option[Int],
                    manpower: Option[Int],
                    maxOrganization: Option[Int],
                    defaultMorale: Option[Int],
                    maxStrength: Option[Int],
                    trainingTime: Option[Int],
                    weight: Option[Double],
                    supplyConsumption: Option[Double]
                  )

object SubUnit {
  /** Read all sub‑unit definitions from files in the given directory. */
  def read(dir: File): List[SubUnit] = {
    if (!dir.isDirectory) throw new IllegalArgumentException(s"$dir is not a directory")

    val files: List[File] = Option(dir.listFiles())
      .map(_.toList.filter(_.isFile))
      .getOrElse(Nil)
    if (files.isEmpty) throw new IllegalArgumentException(s"$dir is empty or has no files")

    val buf = ListBuffer.empty[SubUnit]

    files.foreach { file =>
      val root = try new Parser(file).parse
      catch {
        case e: ParserException =>
          JOptionPane.showMessageDialog(null,
            s"[SubUnit] Error parsing ${file.getName}: ${e.getMessage}")
          return buf.toList   // or simply skip this file
      }
      
      val nodes: Seq[Node] = root.find("sub_units").toSeq
      if (nodes.isEmpty) {
        println(s"No sub_units found in ${file.getName}")
      } else {
        println(s"File: ${file.getName}, subunits: ${nodes.size}")
        buf ++= createSubUnits(nodes)
      }
    }
    buf.toList
  }

  /** Convert parsed nodes into SubUnit instances. */
  private def createSubUnits(nodes: Seq[Node]): Seq[SubUnit] = {
    nodes.map { node =>
      SubUnit(
        node.name,
        if (node.contains("abbreviation")) Some(node.getValue("abbreviation").asString) else None,
        if (node.contains("sprite")) Some(node.getValue("sprite").asString) else None,
        if (node.contains("map_icon_category")) Some(node.getValue("map_icon_category").asString) else None,
        if (node.contains("priority")) Some(node.getValue("priority").$intOrElse(0)) else None,
        if (node.contains("ai_priority")) Some(node.getValue("ai_priority").$intOrElse(0)) else None,
        if (node.contains("active")) Some(node.getValue("active").asBool(BoolType.YES_NO)) else None,
        //subUnit.type//subUnit.type
        if (node.contains("group")) Some(node.getValue("group").asString) else None,
        //subUnit.categories//subUnit.categories
        if (node.contains("combat_width")) Some(node.getValue("combat_width").$intOrElse(0)) else None,
        //subUnit.need//subUnit.need
        if (node.contains("manpower")) Some(node.getValue("manpower").$intOrElse(0)) else None,
        if (node.contains("max_organization")) Some(node.getValue("max_organization").$intOrElse(0)) else None,
        if (node.contains("default_morale")) Some(node.getValue("default_morale").$intOrElse(0)) else None,
        if (node.contains("max_strength")) Some(node.getValue("max_strength").$intOrElse(0)) else None,
        if (node.contains("training_time")) Some(node.getValue("training_time").$intOrElse(0)) else None,
        if (node.contains("weight")) Some(node.getValue("weight").$doubleOrElse(0.0)) else None,
        if (node.contains("supply_consumption")) Some(node.getValue("supply_consumption").$doubleOrElse(0.0)) else None
      )
    }
  }

  /** Data accessor functions, for table display: each SubUnit → a field value. */
  def dataFunctions(): Seq[SubUnit => Any] = Seq(
    _.identifier,
    _.abbreviation,
    _.sprite,
    _.mapIconCategory,
    _.priority,
    _.aiPriority,
    _.active,
    _.group,
    _.combatWidth,
    _.manpower,
    _.maxOrganization,
    _.defaultMorale,
    _.maxStrength,
    _.trainingTime,
    _.weight,
    _.supplyConsumption
  )

  /** Labels corresponding to the dataFunctions order. */
  def dataLabels(): Seq[String] = Seq(
    "Subunit",
    "Abbreviation",
    "Sprite",
    "Map Icon Category",
    "Priority",
    "AI Priority",
    "Active",
    "Group",
    "Combat Width",
    "Manpower",
    "Max Organization",
    "Default Morale",
    "Max Strength",
    "Training Time",
    "Weight",
    "Supply Consumption"
  )
}
